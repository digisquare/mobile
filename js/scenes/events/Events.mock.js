const Events = [{
  Event: {
    id: '1382',
    uid: '230539532@meetup.com',
    edition_id: '9',
    venue_id: '1',
    name: 'Petit retour d’expérience après 6 mois sur une grosse application React',
    description: "Après le talk de Vladimir&nbsp;Barsukov sur Redux, on continue avec React, en accueillant <a href=\"https://twitter.com/damusnet\">Damien Varron</a>, qui nous propose un retour d'expérience sur plusieurs mois de développement.<br><br>React c'est magique, tout le monde en parle, tout le monde vous l'a dit.&nbsp;Mais passé l'exemple du Hello World ou de la TodoMVC comment ça se passe dans la vraie vie ? Est-ce que je m'y retrouve encore dans mon code après plusieurs mois ? Quel est l'intérêt de toutes ces librairies qui fleurissent autour de React ?&nbsp;Voici quelques observations et leçons apprises au cours du développement de l'application <a href=\"https://app.voxeet.com/\">Voxeet</a> et de son <a href=\"http://timezoneninja.com/\">TimeZoneNinja</a>.<br>",
    start_at: '2016-04-28 19:00:00',
    end_at: '2016-04-28 21:00:00',
    status: '0',
    url: 'http://www.meetup.com/BordeauxJS/events/230539532/',
    created: '2016-04-21 15:45:16',
    modified: '2016-04-21 15:45:16',
  },
  Edition: {
    id: '9',
    name: 'Bordeaux',
    slug: 'bordeaux',
    event_count: '906',
    organization_count: '0',
    created: '2015-01-24 13:35:02',
    modified: '2015-01-24 13:35:02',
  },
  Venue: {
    id: '1',
    edition_id: '9',
    name: 'Le Node',
    address: '12 Rue Des Faussets',
    zipcode: '33000',
    city: 'Bordeaux',
    country_code: 'FR',
    latitude: '44.840373',
    longitude: '-0.570311',
    event_count: '277',
    created: '2015-01-24 14:50:42',
    modified: '2015-02-15 17:26:20',
    oneliner: 'Le Node, 12 Rue Des Faussets, 33000 Bordeaux, FR',
  },
  Organization: [
    {
      id: '22',
      venue_id: null,
      edition_id: '9',
      name: 'BordeauxJS',
      type: '0',
      avatar: 'https://pbs.twimg.com/profile_images/2741451195/8ff14ab4849d749af06271795a15cf54_400x400.png',
      description: 'User group JavaScript à #bordeaux : réunions, discussions et live coding entre gens en asynchrone. #nodejs #coffeescript #javascript',
      contacts: '{"twitter":"BordeauxJS","website":"http:\\/\\/www.meetup.com\\/BordeauxJS\\/"}',
      event_count: '7',
      recent_event_count: '0',
      created: '2015-01-24 13:35:25',
      modified: '2016-07-22 04:00:01',
      Contacts: {
        twitter: 'BordeauxJS',
        website: 'http://www.meetup.com/BordeauxJS/',
      },
      Organizer: {
        id: '324',
        event_id: '1382',
        organization_id: '22',
        created: '0000-00-00 00:00:00',
        modified: '0000-00-00 00:00:00',
      },
    },
  ],
}, {
  Event: {
    id: '1420',
    uid: '1420@digisquare.net',
    edition_id: '9',
    venue_id: '7',
    name: 'D&B Talks #28',
    description: "<b>Du design de pages au design de systèmes :</b><br>Plus une interface web grandit, plus il devient difficile de penser son design comme une suite de pages. Nous ne concevons plus des maquettes figées mais des ensembles de composants, amenés à cohabiter sur des appareils de tailles et de résolutions toujours plus variées. Comment créer des styleguides à la fois robustes et modulaires ? Avec quels outils ? Comment améliorent-ils la communication entre les équipes du projet ?<br><br>Ian Smallwood est tombé dans le web par hasard un jour en 1993 alors qu'il cherchait une bibliothèque. Depuis ce jour il essaye de trouver un \"job description\" pour ce qu'il fait. Il a collaboré avec tous (peu-t-être) les profils qu'on peut rencontrer lors d'un projet web. Avec les pattern labs il pense avoir trouvé un endroit qui facilite réellement cette collaboration.<br><br>Ian consacre une partie de son temps à l'enseignement du Web Design et de la culture du web à l'ECV où il apprend beaucoup des questions, idées et besoins de ses étudiants. A Webians c'est en général lui qui répond au téléphone.<br><br>Clem intègre et développe des choses chez Webians. Il est aussi étudiant à l'ECV Digital Bordeaux. Sur son temps libre, il enregistre un album de reprises salsa des plus beaux titres de Motörhead.",
    start_at: '2016-09-26 19:00:00',
    end_at: '2016-09-26 21:00:00',
    status: '0',
    url: 'https://nvite.com/DBTalks28/d3b3',
    created: '2016-09-20 12:47:58',
    modified: '2016-09-20 12:47:58',
  },
  Edition: {
    id: '9',
    name: 'Bordeaux',
    slug: 'bordeaux',
    event_count: '906',
    organization_count: '0',
    created: '2015-01-24 13:35:02',
    modified: '2015-01-24 13:35:02',
  },
  Venue: {
    id: '7',
    edition_id: '9',
    name: "The Ramblin' Man",
    address: '52 Quai Richelieu',
    zipcode: '33000',
    city: 'Bordeaux',
    country_code: 'FR',
    latitude: '44.837321',
    longitude: '-0.566730',
    event_count: '28',
    created: '2015-01-24 18:05:06',
    modified: '2015-01-24 20:58:51',
    oneliner: "The Ramblin' Man, 52 Quai Richelieu, 33000 Bordeaux, FR",
  },
  Organization: [
    {
      id: '15',
      venue_id: '7',
      edition_id: '9',
      name: 'D&B Talks',
      type: '0',
      avatar: 'https://pbs.twimg.com/profile_images/3238741776/26f091b7287359733e08a454f10bbe87_400x400.png',
      description: "D&B Talks c'est le rendez-vous de tous les designers et pros du numérique. On y parle UX, UI, Web, Mobile, motion design...",
      contacts: '{"twitter":"dbtalks_org","facebook":"dbtalksbdx","website":"http:\\/\\/dbtalks.org\\/"}',
      event_count: '20',
      recent_event_count: '0',
      created: '2015-01-24 13:35:25',
      modified: '2016-11-07 04:00:01',
      Contacts: {
        twitter: 'dbtalks_org',
        facebook: 'dbtalksbdx',
        website: 'http://dbtalks.org/',
      },
      Organizer: {
        id: '363',
        event_id: '1420',
        organization_id: '15',
        created: '0000-00-00 00:00:00',
        modified: '0000-00-00 00:00:00',
      },
    },
  ],
}, {
  Event: {
    id: '1409',
    uid: '1409@digisquare.net',
    edition_id: '9',
    venue_id: '11',
    name: 'BDX.IO v3',
    description: 'La journée bordelaise de conférences sur le développement informatique et les technologies de demain.<br><br><a href="https://www.weezevent.com/bdx-i-o-2016">Places en ventes</a> à partir du 16 juin.<br><a href="https://cfp.bdx.io/">Appel à orateur</a> du 8 juin au 24 juillet.',
    start_at: '2016-10-21 09:00:00',
    end_at: '2016-10-21 18:00:00',
    status: '0',
    url: 'http://www.bdx.io/#/',
    created: '2016-06-27 12:10:42',
    modified: '2016-06-27 12:12:31',
  },
  Edition: {
    id: '9',
    name: 'Bordeaux',
    slug: 'bordeaux',
    event_count: '906',
    organization_count: '0',
    created: '2015-01-24 13:35:02',
    modified: '2015-01-24 13:35:02',
  },
  Venue: {
    id: '11',
    edition_id: '9',
    name: 'Enseirb Matmeca',
    address: '1 Avenue Du Dr Albert Schweitzer',
    zipcode: '33400',
    city: 'Talence',
    country_code: 'FR',
    latitude: '44.802614',
    longitude: '-0.588054',
    event_count: '50',
    created: '2015-01-24 18:05:08',
    modified: '2016-05-26 14:03:21',
    oneliner: 'Enseirb Matmeca, 1 Avenue Du Dr Albert Schweitzer, 33400 Talence, FR',
  },
  Organization: [
    {
      id: '35',
      venue_id: '11',
      edition_id: '9',
      name: 'bdx.io',
      type: '0',
      avatar: 'https://pbs.twimg.com/profile_images/465909547752632320/ffD7qZ--_400x400.png',
      description: 'Une journée de conférences sur le développement informatique et les technologies de demain, qui aime bien les puissances de 2 avec près de 512 participants à guichets fermés. La prochaine édition aura lieu le 21 octobre 2016.',
      contacts: '{"twitter":"bdxio","facebook":"bdxio","website":"http:\\/\\/www.bdx.io\\/","societe":""}',
      event_count: '3',
      recent_event_count: '1',
      created: '2015-01-27 21:01:44',
      modified: '2016-11-27 04:00:01',
      Contacts: {
        twitter: 'bdxio',
        facebook: 'bdxio',
        website: 'http://www.bdx.io/',
        societe: '',
      },
      Organizer: {
        id: '351',
        event_id: '1409',
        organization_id: '35',
        created: '0000-00-00 00:00:00',
        modified: '0000-00-00 00:00:00',
      },
    },
  ],
}];

export default Events;
