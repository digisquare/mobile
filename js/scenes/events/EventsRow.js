import React, { PropTypes } from 'react';
import { Text, View, Image, StyleSheet, TouchableOpacity } from 'react-native';
import moment from 'moment/min/moment-with-locales';
import { Answers } from 'react-native-fabric';

import DigiColors from '../../common/DigiColors';

import Event from '../event/Event';

moment.locale('fr');

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    padding: 10,
    height: 60,
  },
  avatarContainer: {
    width: 50,
  },
  avatar: {
    width: 40,
    height: 40,
    marginTop: 3,
  },
  textContainer: {
    flex: 1,
  },
  title: {
    fontSize: 18,
    fontWeight: '900',
    color: DigiColors.secondaryFontColor,
  },
  date: {
    color: DigiColors.secondaryFontColor,
  },
});

const selectEvent = (navigator, event) => {
  Answers.logContentView(
    `${event.Event.name}`,
    'event',
    `events/${event.Event.id}`,
  );
  navigator.push({
    component: Event,
    passProps: {
      event,
    },
  });
};

export default function EventsRow({ rowID, navigator, event }) {
  const time = moment(event.Event.start_at).format('HH:mm');
  const place = event.Venue.name;
  const timeAndPlace = time === '00:00' ? place : `${time} @ ${place}`;
  return (
    <TouchableOpacity
      testID={`EventsRow-${rowID}`}
      onPress={() => selectEvent(navigator, event)}
      style={styles.container}
    >
      <View style={styles.avatarContainer}>
        {
          event.Organization[0] && event.Organization[0].avatar ? (
            <Image
              style={styles.avatar}
              source={{ uri: event.Organization[0].avatar }}
            />
          ) : null
        }
      </View>
      <View style={styles.textContainer}>
        <Text style={styles.title} numberOfLines={1}>
          {event.Event.name}
        </Text>
        <Text style={styles.date}>
          {timeAndPlace}
        </Text>
      </View>
    </TouchableOpacity>
  );
}

EventsRow.propTypes = {
  rowID: PropTypes.string.isRequired,
  navigator: PropTypes.object.isRequired,
  event: PropTypes.object.isRequired,
};
