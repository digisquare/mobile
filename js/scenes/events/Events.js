import React, { Component, PropTypes } from 'react';
import { ListView, RefreshControl, View, Text, StyleSheet } from 'react-native';
import moment from 'moment/min/moment-with-locales';

import DigiError from '../../common/DigiError';
import DigiColors from '../../common/DigiColors';

import EventsRow from './EventsRow';

import tumbeastNetwork from '../../common/img/tumbeast-network.png';
import tumbeastSitting from '../../common/img/tumbeast-sitting.png';

const styles = StyleSheet.create({
  headerContainer: {
    paddingTop: 10,
    paddingBottom: 10,
    borderBottomColor: DigiColors.borderColor,
    borderBottomWidth: 0.5,
    marginLeft: 10,
    marginRight: 10,
    backgroundColor: DigiColors.primaryBackgroundColor,
  },
  header: {
    color: DigiColors.primaryFontColor,
    fontWeight: '900',
    fontSize: 18,
  },
});

export default class Events extends Component {
  constructor(props) {
    super(props);
    this.renderEventsRow = this.renderEventsRow.bind(this);
  }

  renderEventsRow(event, sectionID, rowID) {
    return (
      <EventsRow
        rowID={`${sectionID}-${rowID}`}
        navigator={this.props.navigator}
        event={event}
      />
    );
  }

  render() {
    const { dataSource, error, refreshing, onRefresh } = this.props;
    const refreshControl = (
      <RefreshControl
        refreshing={refreshing}
        onRefresh={onRefresh}
      />
    );

    if (error) {
      return (
        <DigiError
          refreshControl={refreshControl}
          image={tumbeastNetwork}
          text="Oups, impossible de récupérer les évènements"
        />
      );
    }

    if (dataSource.getRowCount() === 0 && !refreshing) {
      return (
        <DigiError
          refreshControl={refreshControl}
          image={tumbeastSitting}
          text="Snif, aucun évènement à venir"
        />
      );
    }

    return (
      <ListView
        testID="EventsListView"
        dataSource={dataSource}
        refreshControl={refreshControl}
        renderSectionHeader={sectionData => (
          <View style={styles.headerContainer}>
            <Text style={styles.header}>
              {moment(sectionData[0].Event.start_at).format('dddd Do MMMM')}
            </Text>
          </View>
        )}
        renderRow={this.renderEventsRow}
      />
    );
  }
}

Events.propTypes = {
  navigator: PropTypes.object.isRequired,
  dataSource: PropTypes.object.isRequired,
  refreshing: PropTypes.bool.isRequired,
  onRefresh: PropTypes.func.isRequired,
  error: PropTypes.bool.isRequired,
};
