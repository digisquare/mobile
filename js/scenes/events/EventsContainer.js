import React, { Component, PropTypes } from 'react';
import { ListView, Linking } from 'react-native';
import { connect } from 'react-redux';
import moment from 'moment/min/moment-with-locales';
import { Answers } from 'react-native-fabric';

import Router from '../../router';

import { fetchEvents } from '../../actions/events';

import Events from './Events';

class EventsContainer extends Component {
  static route = {
    navigationBar: {
      title({ title }) {
        return title || 'Bordeaux';
      },
    },
  }

  constructor(props) {
    super(props);
    const { editions: { selectedEdition }, events } = props;
    const items = events[selectedEdition]
      && events[selectedEdition].items
      && events[selectedEdition].items.length > 0;
    this.state = {
      dataSource: new ListView.DataSource({
        rowHasChanged: (row1, row2) => row1 !== row2,
        sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
      }),
      refreshing: !items,
      error: false,
    };
    this.onRefresh = this.onRefresh.bind(this);
  }

  componentWillMount() {
    const { onFetchEvents, editions, editions: { selectedEdition } } = this.props;
    const editionName = editions.items.find(edition => edition.id === selectedEdition).name;
    onFetchEvents(selectedEdition);
    Answers.logContentView(`Évènements à ${editionName}`, 'events', 'events');
  }

  componentDidMount() {
    const { navigator } = this.props;
    Linking.getInitialURL().then((url) => {
      if (!url) {
        return null;
      }
      const match = url.match(/https:\/\/digisquare.net\/events\/([0-9]+)/);
      if (match && match[1]) {
        navigator.push(Router.getRoute('event', { eventId: parseInt(match[1], 10) }));
      }
      return null;
    }).catch(() => null);
  }

  componentWillReceiveProps(nextProps) {
    const { onFetchEvents, events, editions } = nextProps;
    const { selectedEdition } = editions;
    const { dataSource, refreshing } = this.state;
    if (selectedEdition !== this.props.editions.selectedEdition) {
      const edition = editions.items.find(e => e.id === selectedEdition);
      this.props.navigator.updateCurrentRouteParams({
        title: edition.name,
      });
      onFetchEvents(selectedEdition);
      this.setState({
        refreshing: true,
      });
    }
    if (!events[selectedEdition]) {
      return this.setState({
        dataSource: dataSource.cloneWithRowsAndSections({}),
      });
    }
    const { error, isFetching, items } = events[selectedEdition];
    if (error && !isFetching && items.length === 0) {
      return this.setState({
        refreshing: false,
        error: true,
      });
    }
    if (refreshing && !isFetching) {
      this.setState({
        refreshing: false,
      });
    }
    if (items) {
      const dataBlob = {};
      let date;
      let sectionID = -1;
      events[selectedEdition].items.map((item) => {
        const newDate = moment(item.Event.start_at).format('YYYY-MM-DD');
        if (newDate !== date) {
          date = newDate;
          sectionID += 1;
          dataBlob[sectionID] = [];
        }
        dataBlob[sectionID].push(item);
        return null;
      });
      return this.setState({
        dataSource: dataSource.cloneWithRowsAndSections(dataBlob),
        error: false,
      });
    }
    return null;
  }

  onRefresh() {
    const { onFetchEvents, editions: { selectedEdition } } = this.props;
    onFetchEvents(selectedEdition);
    return this.setState({
      refreshing: true,
    });
  }

  render() {
    const { navigator } = this.props;
    const { dataSource, refreshing, error } = this.state;
    return (
      <Events
        navigator={navigator}
        dataSource={dataSource}
        refreshing={refreshing}
        onRefresh={this.onRefresh}
        error={error}
      />
    );
  }
}

const mapStateToProps = state => ({
  editions: state.editions,
  events: state.events,
});

const mapDispatchToProps = dispatch => ({
  onFetchEvents: (selectedEdition) => {
    dispatch(fetchEvents(selectedEdition));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(EventsContainer);

EventsContainer.propTypes = {
  navigator: PropTypes.object.isRequired,
  onFetchEvents: PropTypes.func.isRequired,
  editions: PropTypes.object.isRequired,
  events: PropTypes.object.isRequired,
};
