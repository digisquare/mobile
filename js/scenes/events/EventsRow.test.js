import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';

import EventsRow from './EventsRow';

import events from './Events.mock';

jest.mock('react-native-fabric', () => ({
  crash: () => {},
}));

jest.mock('@exponent/ex-navigation', () => ({
  withNavigation: () => {},
  createRouter: () => {},
}));

it('renders correctly', () => {
  const tree = renderer.create(
    <EventsRow rowID="0-0" navigator={{ push: () => null }} event={events[0]} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
