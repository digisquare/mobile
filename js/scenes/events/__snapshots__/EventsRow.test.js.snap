exports[`test renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "height": 60,
      "padding": 10,
    }
  }
  testID="EventsRow-0-0">
  <View
    style={
      Object {
        "width": 50,
      }
    }>
    <Image
      source={
        Object {
          "uri": "https://pbs.twimg.com/profile_images/2741451195/8ff14ab4849d749af06271795a15cf54_400x400.png",
        }
      }
      style={
        Object {
          "height": 40,
          "marginTop": 3,
          "width": 40,
        }
      } />
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "grey",
          "fontSize": 18,
          "fontWeight": "900",
        }
      }>
      Petit retour d’expérience après 6 mois sur une grosse application React
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "grey",
        }
      }>
      19:00 @ Le Node
    </Text>
  </View>
</TouchableOpacity>
`;
